agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -4382,16 +4382,66 @@ bool slave_thread::handle_channel_preference_query(Socket *sd, ieee1905_1::CmduM
+    for (auto preference : channel_preferences) {
+        LOG(DEBUG) << "Preference operating class: " << int(preference.preference.oper_class);
+        for (uint8_t i = 0; i < beerocks::message::SUPPORTED_CHANNELS_LENGTH; i++) {
+            auto channel         = hostap_params.supported_channels[i];
+            auto operating_class = wireless_utils::get_operating_class_by_channel(
+                channel.channel, static_cast<beerocks::eWiFiBandwidth>(channel.channel_bandwidth));
+
+            // Skip channels from other operating classes
+            if (operating_class != preference.preference.oper_class) {
+                continue;
+            }
+            // Skip DFS channels
+            if (channel.is_dfs_channel) {
+                LOG(DEBUG) << "Skip DFS channel " << int(channel.channel);
+                continue;
+            }
+            auto is_restricted =
+                std::find_if(preference.channels.begin(), preference.channels.end(),
+                             [&](const beerocks::message::sWifiChannel &wifi_channel) {
+                                 return (channel.channel == wifi_channel.channel);
+                             });
+            // Skip restricted channels
+            if (is_restricted != preference.channels.end()) {
+                LOG(DEBUG) << "Skip restricted channel " << int(channel.channel);
+                continue;
+            }
+            // If we got this far, we found a candidate channel, so switch to it
+            LOG(DEBUG) << "Selected channel " << int(channel.channel);
+            return channel;
+        }
+    }
 
-    bool power_limit_received = false;
-    int switch_required       = 0;
-    int power_limit           = 0;
-    beerocks::message::sWifiChannel channel_to_switch;
+    LOG(ERROR) << "Could not find a suitable channel";
+    return beerocks::message::sWifiChannel();
+}
+
+bool slave_thread::channel_selection_current_channel_restricted()
+{
+    for (auto preference : channel_preferences) {
+        auto is_restricted =
+            std::find_if(preference.channels.begin(), preference.channels.end(),
+                         [&](const beerocks::message::sWifiChannel &wifi_channel) {
+                             return hostap_cs_params.channel == wifi_channel.channel;
+                         });
 
+        if (is_restricted != preference.channels.end()) {
+            LOG(INFO) << "Current channel " << int(hostap_cs_params.channel)
+                      << " is restricted, channel switch required";
+            return true;
+        }
+    }
+    LOG(INFO) << "Current channel " << int(hostap_cs_params.channel)
+              << " not restricted, channel switch not required";
+    return false;
+}
+
+bool slave_thread::channel_selection_get_channel_preference(ieee1905_1::CmduMessageRx &cmdu_rx)
+{
     channel_preferences.clear();
 
     for (auto channel_preference_tlv : cmdu_rx.getClassList<wfa_map::tlvChannelPreference>()) {
```

Comment:
You should use 'const auto&'
I think `cmdu_rx` parameter should also be a const reference

Reply:
This one has been marked as resolved but it is not.
This file deserves a full review regarding the 'const auto&' usage.